webob/response.py

```diff
@@ -225,43 +222,66 @@ def __init__(self, body=None, status=None, headerlist=None, app_iter=None,
         # Response with empty body, such as Response(status='204 No Content')
         # without the default content_type being set
 
-        if (
-            self.content_type is None and
-            headerlist is None and
-            _code_has_body(self.status_code)
-        ):
-            self.content_type = content_type
+        encoding = None
+        code_has_body = (
+            self._status[0] != '1' and
+            self._status[:3] not in ('204', '205', '304')
+        )
+        if headerlist is None and code_has_body:
+            content_type = content_type or self.default_content_type
 
-        # Set up the charset
-        #
-        # In contrast with the above, if a charset is not set but there is a
-        # content_type we will set the default charset if the content_type
-        # allows for a charset.
+            # Check if content_type is set because default_content_type could
+            # be None, in which case there is no content_type
+            if content_type:
+
+                # Set up the charset, if the content_type doesn't already have one
 
-        if self.content_type:
-            if not self.charset and charset is not _marker:
-                self.charset = charset
-            elif not self.charset and self.default_charset:
-                if _content_type_has_charset(self.content_type):
-                    self.charset = self.default_charset
+                has_charset = 'charset=' in content_type
+
+                if not has_charset and charset is not _marker and charset is not None:
+                    encoding = charset
+                    content_type += '; charset=' + charset
+                elif not has_charset and charset is _marker and self.default_charset:
+                    # Optimize for the default_content_type case, avoid a function call
+                    if (
+                        content_type is self.default_content_type or
+                        _content_type_has_charset(content_type)
+                    ):
```

Comment:
What about baking the charset into the `default_content_type` instead of handling them separately in this case? Then `default_content_type` is unambiguous, and `default_charset` is only used if someone provides a partial content-type.